numpy/polynomial/chebyshev.py

```diff
@@ -1766,6 +1936,255 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+
+    Notes
+    -----
+    The solution is the coefficients of the Chebyshev series `p` that
+    minimizes the sum of the weighted squared errors
+    .. math:: E = \\sum_j w_j^2 * |y_j - p(x_j)|^2,
+    where :math:`w_j` are the weights. This problem is solved by setting up
+    as the (typically) overdetermined matrix equation
+    .. math:: V(x, y) * c = w * y,
+    where `V` is the weighted pseudo Vandermonde matrix of `x`, `c` are the
+    coefficients to be solved for, `w` are the weights, and `y` are the
+    observed values.  This equation is then solved using the singular value
+    decomposition of `V`.
+    If some of the singular values of `V` are so small that they are
+    neglected, then a `RankWarning` will be issued. This means that the
+    coefficient values may be poorly determined. Using a lower order fit
+    will usually get rid of the warning.  The `rcond` parameter can also be
+    set to a value smaller than its default, but the resulting fit may be
+    spurious and have large contributions from roundoff error.
+    Fits using Chebyshev series are usually better conditioned than fits
+    using power series, but much can depend on the distribution of the
+    sample points and the smoothness of the data. If the quality of the fit
+    is inadequate splines may be a good alternative.
+
+    References
+    ----------
+    .. [1] Wikipedia, "Curve fitting",
+           http://en.wikipedia.org/wiki/Curve_fitting
+    Examples
+    --------
+    """
+    xi_ = np.array(xi, copy=0) + 0.0
+    z = np.array(f)
+    degrees = np.asarray(deg, dtype=int)
+    orders = degrees + 1
+    order = np.product(orders)
+
+    ndims = np.array([x.ndim for x in xi_])
+    ndim = len(ndims)
+    sizes = np.array([x.size for x in xi_])
+    if np.any(ndims!=ndim) or z.ndim!=ndim:
+        raise TypeError("expected %dD array for x1, x2,...,xn and f" % ndim)
+    if np.any(sizes == 0):
+        raise TypeError("expected non-empty vector for xi")
+
+    lhs = chebvandernd(degrees, *xi).reshape((-1, order))
+    rhs = z.ravel()
+    if w is not None:
+        w = np.asarray(w).ravel() + 0.0
+        if len(lhs) != len(w):
+            raise TypeError("expected x and w to have same length")
+        lhs = lhs * w
+        rhs = rhs * w
+
+    if rcond is None:
+        rcond = xi_[0].size * np.finfo(xi_.dtype).eps
+
+    if issubclass(lhs.dtype.type, np.complexfloating):
+        scl = np.sqrt((np.square(lhs.real) + np.square(lhs.imag)).sum(axis=0))
+    else:
+        scl = np.sqrt(np.square(lhs).sum(axis=0))
+    scl[scl == 0] = 1
+
+    # Solve the least squares problem.
+    c, resids, rank, s = np.linalg.lstsq(lhs/scl, rhs, rcond)
+    c = (c/scl).reshape(orders)
+
+    if rank != order and not full:
+        msg = "The fit may be poorly conditioned"
+        warnings.warn(msg, pu.RankWarning)
+
+    if full:
+        return c, [resids, rank, s, rcond]
+    else:
+        return c
+
+
+# def chebfit_dct(f, n=(10, ), domain=None):
```

Comment:
The name is too method specific, i.e., `dct`. Maybe `chebfitfunc` or some such. You can use the Gauss points here with the weights, a method which is probably not much slower than the dct for arbitrary values of `n` that don't have nice factorizations and will generalize to polynomials other than Chebyshev.